cabal-install/src/Distribution/Client/Init/Command.hs

```diff
@@ -208,29 +210,34 @@ getCabalVersion flags = do
 -- | Get the package name: use the package directory (supplied, or the current
 --   directory by default) as a guess. It looks at the SourcePackageDb to avoid
 --   using an existing package name.
-getPackageName :: SourcePackageDb -> InitFlags -> IO InitFlags
-getPackageName sourcePkgDb flags = do
-  guess    <-     traverse guessPackageName (flagToMaybe $ packageDir flags)
-              ?>> Just `fmap` (getCurrentDirectory >>= guessPackageName)
-
-  let guess' | isPkgRegistered guess = Nothing
-             | otherwise = guess
-
-  pkgName' <-     return (flagToMaybe $ packageName flags)
-              ?>> maybePrompt flags (prompt "Package name" guess')
-              ?>> return guess'
-
-  chooseAgain <- if isPkgRegistered pkgName'
-                    then promptYesNo promptOtherNameMsg (Just True)
-                    else return False
+getPackageName :: Verbosity -> SourcePackageDb -> InitFlags -> IO InitFlags
+getPackageName verbosity sourcePkgDb flags = do
+  guess <- maybe (getCurrentDirectory >>= guessPackageName) pure
+             =<< traverse guessPackageName (flagToMaybe $ packageDir flags)
+
+  pkgName' <- case flagToMaybe $ packageName flags of
+    Just pkgName -> return $ Just $ pkgName
+    _ -> maybePrompt flags (prompt "Package name" (Just guess))
+  let pkgName = fromMaybe guess pkgName'
+
+  chooseAgain <- if isPkgRegistered pkgName
+                   then do
+                     answer' <- maybePrompt flags (promptYesNo promptOtherNameMsg (Just True))
+                     case answer' of
+                       Just answer -> return answer
+                       _ -> die' verbosity $ deathNameMsg pkgName
+                 else
+                   return False
 
   if chooseAgain
-    then getPackageName sourcePkgDb flags
-    else return $ flags { packageName = maybeToFlag pkgName' }
+    then getPackageName verbosity sourcePkgDb flags
+    else return $ flags { packageName = Flag pkgName }
 
   where
-    isPkgRegistered (Just pkg) = elemByPackageName (packageIndex sourcePkgDb) pkg
-    isPkgRegistered Nothing    = False
+    isPkgRegistered pkg = elemByPackageName (packageIndex sourcePkgDb) pkg
+
+    deathNameMsg pkgName = "The name " ++ (P.unPackageName pkgName) ++
```

Comment:
Could you merge this function (which takes the package name for a more descriptive message) with `prompOtherNameMsg` and use the latter in both places?